Mathlib/Order/Quantale.lean

```diff
@@ -0,0 +1,128 @@
+* `x ⇨ᵣ y` : `sSup { z | x * z ≤ y }`, the right-residuation of `y` over `x`;
+
+## References
+
+[Topology via Logic][vickers1989]
+<https://en.wikipedia.org/wiki/Quantale>
+<https://encyclopediaofmath.org/wiki/Quantale>
+<https://ncatlab.org/nlab/show/quantale> (Categorical definition - not followed here)
+
+## TODO
+
+Additive quantales and multiplicative quantales both occur in literature.
+For sequences of actions, usually a multiplicative quantale is used, but for
+describing timing of actions (for example) the additive quantale of max-plus algebra
+a.k.a. tropical algebra, can be used. Therefore, extending the definitions with
+addition is - I think - relevant future work.
+
+-/
+
+/-- A quantale is a monoid distributing over a complete sup-semilattice.
+
+    Technically, I would prefer to define it as
+    `class Quantale (α : Type*) extends Monoid α, CompleteSemilatticeSup α where ...`
+    however, if I do that currently, the elements ⊤ and ⊥ are not automatically defined.
+    I could do that separately, but it seems more sensible to push the definition
+    of those to CompleteSemmilatticeSup in the library, as they are already making
+    sense there and having them should not break anything (in theory).
+
+    The reason not to consider a quantale as a complete lattice, is because the
+    morphisms between quantales usually only preserve the join and not the inf.
+    In fact, when seen as a generalization of locales, the monoid multiplication
+    usually takes the place of the inf.
+-/
+class Quantale (α : Type*) extends Monoid α, CompleteLattice α where
+  /-- Multiplication is distributive over join in a quantale -/
+  protected mul_sSup_eq_iSup_mul (x : α) (s : Set α) : x * sSup s = ⨆ y ∈ s, x * y
+  /-- Multiplication is distributive over join in a quantale -/
+  protected sSup_mul_eq_iSup_mul (s : Set α) (y : α) : sSup s * y = ⨆ x ∈ s, x * y
+
+section
+
+variable (α : Type _)
+variable [Quantale α]
+
+theorem mul_sSup_eq_iSup_mul : ∀ x : α, ∀ s : Set α, x * sSup s = ⨆ y ∈ s, x * y :=
+  Quantale.mul_sSup_eq_iSup_mul
+theorem sSup_mul_eq_iSup_mul : ∀ s : Set α, ∀ y : α, sSup s * y = ⨆ x ∈ s, x * y :=
+  Quantale.sSup_mul_eq_iSup_mul
+
+namespace Quantale
+
+/- Basic definitions on quantales -/
+
+/-- A quantale is integral if its unit and top element coincide -/
+def is_integral [Quantale α] := (⊤ : α) = 1
+
+/-- A quantale is commutative if its monoid satisfies `x * y = y * x` -/
+def is_comm [Quantale α] := ∀ x y : α , x * y = y * x
+
+/-- A quantale is idempotent if its monoid satisfies `x * x = x` -/
+def is_idem [Quantale α] := ∀ x : α, x * x = x
+
+variable {α : Type _}
+variable [Quantale α]
+
+/-- Left- and right- residuation operators on a quantale are similar to the Heyting operator
+    on complete lattices, but for a non-commutative logic.
+    I.e. `x ⇨ₗ y = sSup { z | z * x ≤ y }`.
+-/
+def l_residuation (x y : α) := sSup { z | z * x ≤ y }
+/-- Notation for left-residuation in quantales.
```

Comment:
style, we generally leave gaps between declarations, and I would probably just use the word `left` instead of abbreviating it in the name. This name is not too long that it needs abbreviations.
```suggestion
def left_residuation (x y : α) := sSup { z | z * x ≤ y }

/-- Notation for left-residuation in quantales.
```